File: viz/svg.mjs
// <include file="machine/comments.xml" path="//term[@id='svg.mjs']"/>
// Graphviz 가 낸 SVG 를 HTML 본문에 그대로 넣을 수 있게 다듬는 스크립트.
// 쓰는 것: 없음 · 쓰이는 곳: 없음
// viz/svg.mjs
// Graphviz 가 낸 SVG 를 HTML 본문에 인라인할 수 있는 형태로 정규화한다.
// 규칙: 헤더 제거 / width·height 제거하고 viewBox 유지 / id 접두사.

const PT_TO_PX = 4 / 3;

/**
 * @param {string} raw       dot -Tsvg_inline 출력
 * @param {string} idPrefix  한 페이지에 SVG 가 2개 이상일 때의 충돌 방지 접두사
 * @returns {{svg: string, naturalWidthPx: number|null, naturalHeightPx: number|null}}
 */
export function inlineSvg(raw, idPrefix) {
  const start = raw.indexOf("<svg");
  if (start === -1) throw new Error("<svg> 엘리먼트를 찾지 못했다");

  const rest = raw.slice(start);
  const tagEnd = rest.indexOf(">");
  let openTag = rest.slice(0, tagEnd + 1);
  const body = rest.slice(tagEnd + 1);

  const w = openTag.match(/\bwidth="([\d.]+)pt"/);
  const h = openTag.match(/\bheight="([\d.]+)pt"/);
  const naturalWidthPx = w ? Math.round(Number(w[1]) * PT_TO_PX) : null;
  const naturalHeightPx = h ? Math.round(Number(h[1]) * PT_TO_PX) : null;

  // width/height 제거. viewBox 는 그대로 둔다 — 반응형 축소는 CSS 가 맡는다.
  openTag = openTag.replace(/\s+(?:width|height)="[^"]*"/g, "");

  let svg = openTag + body;

  // id 정의와 그 참조 3종을 함께 치환한다. 하나라도 빠지면 clipPath/marker 가 깨진다.
  svg = svg.replace(/\bid="([^"]+)"/g, (_, id) => `id="${idPrefix}-${id}"`);
  svg = svg.replace(/url\(#([^)]+)\)/g, (_, id) => `url(#${idPrefix}-${id})`);
  svg = svg.replace(/\b(xlink:href|href)="#([^"]+)"/g, (_, attr, id) => `${attr}="#${idPrefix}-${id}"`);

  return { svg, naturalWidthPx, naturalHeightPx };
}

File: viz/svg.py
import re
from typing import Dict

def inlineSvg(svg: str, props: Dict[str, str]) -> str:
    m = re.search(r'<svg([^>]*)>', svg)
    if not m:
        return svg
    attrs_str = m.group(1)
    
    attr_dict: Dict[str, str] = {}
    for attr_match in re.finditer(r'([A-Za-z0-9-]+)="([^"]*)"', attrs_str):
        attr_dict[attr_match.group(1)] = attr_match.group(2)
        
    for k, v in props.items():
        if v == "" and k in attr_dict:
            del attr_dict[k]
        elif v != "":
            attr_dict[k] = v
            
    if "width" in attr_dict and "height" in attr_dict and "viewBox" not in attr_dict:
        w = attr_dict["width"].replace("pt", "")
        h = attr_dict["height"].replace("pt", "")
        attr_dict["viewBox"] = f"0 0 {w} {h}"
        
    new_attrs = " ".join(f'{k}="{v}"' for k, v in attr_dict.items())
    return f"<svg {new_attrs}>" + svg[m.end():]
